### PR TITLE
Add navigation menu blocks

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR adds navigation menu blocks it is a first small step to get https://github.com/WordPress/gutenberg/issues/13690.
It is a first small interaction using only what blocks can do right now.
It still contains the nested blocks UI.
But it creates a basis from where we can follow up in parallel.

Short term follow up tasks to get to the v1:
- Finish the inserter component as shown in https://github.com/WordPress/gutenberg/issues/13690.
- Add a logic that when a navigation menu is created we automatically generate a block list based on top level pages.
- Allow InnerBlocks to be customized so we can remove the wrapping block UI.
- Expose block Drag & Drop/mover functionality so it can be customized/reused here.

Changes outside block-library done here are small improvements to make this possible but they are not related to this PR and will be extracted and properly documented.

## How has this been tested?
I checked it is possible to create a menu, I checked the output markup is the expected one.

### CODE REVIEW

File: packages/block-library/src/index.js
Comment (gziolo) [inline packages/block-library/src/index.js]: `navigationMenuItem` will have to be moved to its own folder to make it work with `block.json`  metadata file. I think it's fine to put it in the root folder.

**Review by gziolo — COMMENTED** [review]
It would be nice to follow https://github.com/WordPress/gutenberg/pull/14770 and introduce `block.json` file for new blocks.

File: packages/block-library/src/index.js
Comment (gziolo) [reply packages/block-library/src/index.js]: Thanks 👍

File: packages/block-editor/src/components/plain-text/index.js
Comment (gziolo) [inline packages/block-editor/src/components/plain-text/index.js]: Should it be handled in its own PR?

File: packages/block-library/src/navigation-menu/editor.scss
Comment (gziolo) [inline packages/block-library/src/navigation-menu/editor.scss]: Should it be extracted to the file specific to the Navigation Menu Item block?

File: packages/block-library/src/navigation-menu/menu-item-inserter.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu/menu-item-inserter.js]: You can experiment with `useState` and `useCallback` React hooks :)

Optional, I read https://overreacted.io/a-complete-guide-to-useeffect/ today from @gaearon and got inspired 😃

File: packages/block-library/src/navigation-menu/edit.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu/edit.js]: It can be expressed as a function component.

File: packages/block-library/src/navigation-menu-item/edit.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu-item/edit.js]: This file has already 250 lines. I'm wondering if `MenuItemActions` should be extracted to its file.

File: packages/block-editor/src/components/plain-text/index.js
Comment (gziolo) [reply packages/block-editor/src/components/plain-text/index.js]: I see it now: #14866 👍

**Review by gziolo — COMMENTED** [review]
How about we put this block behind the `GUTENBERG_PHASE` feature flag and merge it later this week? @mapk how does it look behavior wise? 

_Aside: Reviewing 750 lines long PR is going to be suboptimal if we continue iterating too long on this PR._

**Comment by jorgefilipecosta** [comment]
Hi @gziolo all your feedback was applied. I was in doubt if I should use hooks given that we are not using them widely on our code base. But I think it is time to start and given that we are implementing totally new blocks it seems a nice opportunity. I refactored the code to use hooks and simplified it.

File: packages/block-library/src/navigation-menu-item/index.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu-item/index.js]: The direction is to have all client-side specific cod in its own file. `save` should be moved to `save.js` file. This will allow to get rid of `index.js` completely at some point.

File: packages/block-library/src/navigation-menu/index.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu/index.js]: The same note about `save` as for another block :)

File: packages/block-library/src/navigation-menu-item/index.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu-item/index.js]: It looks like it can be removed.

File: packages/block-library/src/navigation-menu/menu-item-inserter.js
Comment (gziolo) [inline packages/block-library/src/navigation-menu/menu-item-inserter.js]: Will ever `insertMenuItem` change? Well, it's a prop passed down so React team probably recommends to include it. No action required, it's just me trying to better understand how those hooks should work :)

File: packages/block-library/src/index.js
Comment (gziolo) [inline packages/block-library/src/index.js]: Can you put those blocks behind the feature flag similar to Legacy Widget?

**Review by gziolo — COMMENTED** [review]
Code wise, it looks great. I love how those React hooks make everything simpler. After all, it isn't that much to learn after you do the homework by reading their docs and the article I referenced in my previous comment. 

Love your latest changes ❤️

**Comment by jorgefilipecosta** [comment]
Thank you for the code review @gziolo! The feedback was applied I think from the code point of view this getting ready.

From the design/UX point of view, we are not there yet, getting there would create an even bigger PR making it impossible to review in one go so we will need some follow up PR's.
For reference here are some images of the current behavior:


Apr-12-2019 12-29-48

Apr-12-2019 12-32-54
(ends with using the keyboard to navigate around all the items).


I guess there are two options:
We merge this PR with the blocks exposed as they are now (behind the phase 2 flag so they appear only on the plugin) marked as experimental.
We merge but we add an {inserter: false } flag making it impossible for anyone to notice these blocks. E.g: they don't appear anywhere unless the flag is changed.


I guess we should go with option 1 if we think these samples already provides some value and may be useful and we want testing to be easy.
We should go with option 2 if we think we are still in a very early stage and we prefer to keep the blocks hidden, making people that want to test the block need to change the flag.

Any thoughts on this @mapk, @melchoyce, @sarahmonster?

**Comment by melchoyce** [comment]
> We merge this PR with the blocks exposed as they are now (behind the phase 2 flag so they appear only on the plugin) marked as experimental.

Would they be available for anyone using the plugin after the next release? If so, I think I'd opt for hiding them for now.

**Comment by sarahmonster** [comment]
I'd prefer to opt for option #2 until we have our MVP of a v1 up and running, as per [this comment](https://github.com/WordPress/gutenberg/issues/13690#issuecomment-480102173) (scroll waaaaaay down!). Let's be very deliberate with how we reveal this functionality to users.

**Comment by jorgefilipecosta** [comment]
I hided the block from the inserter. It is possible to see it by executing the following code in the browser console:
```
(() => { const navBlock = wp.data.select('core/blocks').getBlockType('core/navigation-menu'); wp.data.dispatch('core/blocks').addBlockTypes( { ...navBlock, supports: { ...navBlock.supports, inserter: true } } ); })();
```

**Comment by sarahmonster** [comment]
I've been trying to test this PR but can't seem to get the navigation menu block to show up, either under "Layout Elements" or by using the slash inserter. @jorgefilipecosta, can you clarify how to test this to see it working?

**Comment by sarahmonster** [comment]
Hm.

> TypeError: navBlock is undefined

I might be doing it wrong or need to configure something additional?

**Comment by jorgefilipecosta** [comment]
Hi @sarahmonster I'm sorry about the problems you are having. I separated the code in lines to make it easier to paste if I totally paste the code block I'm not seeing the error, if possible give another try to the new version.
Another way to be able to test the block would be pasting a block already created in the code editor:
```
<!-- wp:navigation-menu -->
<ul class="wp-block-navigation-menu"><!-- wp:navigation-menu-item {"label":"WordPress","destination":"http://wordpress.org/"} -->
<li class="wp-block-navigation-menu-item"><a href="http://wordpress.org/">WordPress</a></li>
<!-- /wp:navigation-menu-item --></ul>
<!-- /wp:navigation-menu -->
```


When switching to the visual editor the block should be there.

Both methods only work on this branch.

**Comment by sarahmonster** [comment]
It finally worked! 🎊 Turns out my whole Docker setup was borked and just needed to run endlessly and turn my computer into a flight device in order to make it all work.

I have a few notes after testing out the PR, but none of them are necessarily critical to address right now, since things are still very much in development. It's really helpful seeing this in context of the current editor. Feel free to ignore for now and we can address them at a later iteration! 

- "Edit label text" currently just moves focus, but doesn't close the drop-down. It might be more clear what's happening if we close the drop-down.
- When the menu is output, should it use a `<ul>` element, or [would a simple `<nav>` work better](https://css-tricks.com/navigation-in-lists-to-be-or-not-to-be/)? I'm happy to optimise for accessibility of content here.
- The link adding insert mechanism could be smoothed out a bit, but let's tackle this once we have the add/search interface up and running, since that's going to be the primary use case here.
- When the menu item is selected, there's no indication that it's part of a larger menu block. We'll want to have block outlines for the parent block so as to indicate that you can click out into the parent.

I think one thing we may want to solidify a bit here is how the menu items themselves look & appear. There are two approaches here: using just a single drop-down, or using a simplified version of the block toolbar. Both have different advantages. 

Right now, this PR does both. :) We should pick a direction and follow it—although I don't think we necessarily have to settle on one direction just yet. Here's a visual comparison, along with pros/cons of each approach:

## Block toolbar



Pros:
- common pattern
- allows for more extensibility, if needed
- easy to access additional tools like "insert between"

Cons:
- strong left border feels awkward with the horizontal layout and cuts off the block mover a bit
- feels a bit more "heavy"
- we might need to use a new icon here for move controls, rather than lump them under the ellipsis menu, since the ellipsis menu is pretty consistent in terms of what's inside it across different blocks


## Custom drop-down



Pros:
- tested really well with users (the dropdown was suuuuuuper intuitive for people, regardless of familiarity with Gutenberg)
- makes sense from an accessibility perspective (the tab order is clear)
- more compact & less visual distraction

Cons:
- introduces a slightly new pattern for users to learn
- could end up getting quite long/complex if we end up needing to add more controls.

@LukePettway, would you be able to weigh in from an accessibility perspective here?

**Comment by LukePettway** [comment]
@sarahmonster Ran into the same issues you did getting this running but I can test it out now.  As far as the two different designs go, I do think I personally prefer the second because the logical flow matches the document order of all the controls. 

I definitely agree that the dropdown could become crowded overtime. Maybe having two controls to the right of the input, one for adjusting position, and one for settings might help solve that? As it stands, most of the items in the dropdown are related to positioning menu items. 

**Some Observations from testing with Voice Over:**

One thing I just noticed is that once the navigation menu block gets focus, the next tab takes you to the add block button:

focus-add-block

Update: I get it now, this is the add new block button for adding a menu item to the navigation. The labeling on this might need some thinking through to make it a little more contextual.

When using VoiceOver and navigating through the dropdown controls menu gets repeated due to the label + hows VO reads out menu items in a dropdown menu. 

Example -> (Remove from menu, menu item)

**Comment by jorgefilipecosta** [comment]
Hi @sarahmonster, @LukePettway thank you for the feedback.

> "Edit label text" currently just moves focus, but doesn't close the drop-down. It might be more clear what's happening if we close the drop-down.

Fixed.

> When the menu is output, should it use a <ul> element, or would a simple <nav> work better? I'm happy to optimise for accessibility of content here.

I used <ul> because it seemed current menus also used it. I updated the code to use nav as suggested.

> When the menu item is selected, there's no indication that it's part of a larger menu block. We'll want to have block outlines for the parent block so as to indicate that you can click out into the parent.

I agree but it seems like a problem that affects any type of nested blocks right? When a child is selected, there is no indication that it is part of a large block. I guess maybe we can add an option for blocks to keep outlines if one of its childs is selected, so other blocks in the community can benefit from this? If that's the case we should implement this outside of the scope of this PR as it is a general blocks/InnerBlocks feature.





>I think one thing we may want to solidify a bit here is how the menu items themselves look & appear. There are two approaches here: using just a single drop-down, or using a simplified version of the block toolbar. Both have different advantages.

> Right now, this PR does both. :) We should pick a direction and follow it—although I don't think we necessarily have to settle on one direction just yet. Here's a visual comparison, along with pros/cons of each approach


Right now I'm trying to follow the "Custom drop-down". The ellipsis menu still appears because currently there is no option to remove the menu/Block toolbar, every block contains it. If we follow the "Custom drop-down" path we will need to add a feature that allows blocks to not have an Ellipsis menu or more generally the Block toolbar.
If we follow the option of using the Block toolbar, we will need to allow blocks to customize the items that appear on the ellipsis menu, right now blocks have no control over what appears there.
Both options will require us to add a new extensibility mechanism.
To make it more clear that this implementation was trying to follow "Custom drop-down" path I remove the toolbar with CSS. I guess we can keep this "fix" for testing purposes until we have a proper way to do it.


> One thing I just noticed is that once the navigation menu block gets focus, the next tab takes you to the add block button:

focus-add-block

This button should probably not appear in this context as we have a specific inserter for menu items. I remove it with CSS for now, until we have a proper solution to remove the sibling inserter from InnerBlocks.


> When using VoiceOver and navigating through the dropdown controls menu gets repeated due to the label + hows VO reads out menu items in a dropdown menu. 

> Example -> (Remove from menu, menu item)


Inter

**Comment by sarahmonster** [comment]
It sounds like, from an accessibility perspective, the custom drop-down approach is the best one. It's the one I'm currently leaning toward as well, but I'd love to get some additional feedback here, so I'm adding the design feedback label. Once we have a clear direction, we should be able to push this forward.

I  agree with you @jorgefilipecosta that we'll want to address the parent block borders in a separate PR. (I think there may be work happening there elsewhere, as well.)

**Comment by kjellr** [comment]
Regarding the block toolbar versus custom dropdown approaches, I agree that the custom dropdown seems to work well. I'm generally against creating new patterns when patterns already exist, but in practice, the custom dropdown feels pretty natural to me. The actions in the ellipsis menu are seldom used (that's why they're framed as overflow from an icon perspective), but the icons for rearranging these menu items are going to be used pretty often. If we do continue with the custom dropdown approach, we should be prepared to write up some guidelines on why/how/when that pattern should be used in other blocks. 

One small additional note: I do worry whether it'll get confusing to have the arrow dropdown there once we introduce nested menu items — I know we represented that in testing, but it still feels unideal.  But we can deal with that when we get to it. 🙂

**Comment by sarahmonster** [comment]
> One small additional note: I do worry whether it'll get confusing to have the arrow dropdown there once we introduce nested menu items — I know we represented that in testing, but it still feels unideal. But we can deal with that when we get to it. 🙂

Yeah, this one really surprised me when we tested it, because I expected to see people getting confused—but they just didn't, somehow. 

We did end up using slightly different arrows, which seemed to help:

Unselected:
Screenshot 2019-05-15 at 14 08 23

Selected:
Screenshot 2019-05-15 at 14 05 29

I'm not a huge fan of the double-arrow that ends up appearing here, but it seems to be the best way to communicate both "open for more controls" and also "I am a sub-menu, hear me roar." It would definitely be worth paying close attention to this for future usability testing, to be sure it doesn't trip people up. There are .... an awful lots of arrows in a very small space here.

**Comment by mapk** [comment]
I will note that while testing this just now, after adding an additional menu item, I then wanted to remove it. I anxiously searched around for the ellipses menu to remove the block. Finally I realized that these actions had been moved to the arrow button, but yet still had a quick panic about this.

I'm very used to the other pattern at this point, but I can see becoming familiar with the new pattern as well in time.

**Comment by mapk** [comment]
I came back to this again and noticed something that I thought might be good to discuss. The way this is built right now shows that the dropdown arrow which resides _inside_ the block contains the actions that controls the block itself. I'm not entirely sure if this is because the design isn't fully implemented, but it does seem a bit odd that a button _inside_ the block controls the block. Here's an example below.

nav

If this is because the design isn't fully implemented just ignore me. 😉

**Comment by sarahmonster** [comment]
@mapk Yes, that's because it's half-implemented and still has the block borders—it should look like this instead:

Screenshot 2019-05-15 at 14 05 29

Since these are big changes, we're making them in small steps in the interests of keeping PRs small. That's another part of why the block is entirely hidden from users. :)

@jorgefilipecosta, I think it would be good to merge this so we can start working on the other pieces of the puzzle. Do we need a last developer review prior to merging?

**Comment by jorgefilipecosta** [comment]
Hi @sarahmonster, given the number of additions in this PR I guess we would benefit from a last code review by a developer.

**Comment by sarahmonster** [comment]
👍 I've gone ahead and requested some re-reviewing, and if we need extra help we can surface in this weeks' core-editor chat. Then we can move on to the next stages! Sound good?

**Comment by earnjam** [comment]
I haven't tested this out yet, but on a quick pass my main concern is this completely changes the way nav menus are stored in the database. That's going to be a concern for backward compatibility. 

Currently nav menus are stored as a taxonomy term, with each item stored as a post and assigned that term. I'm not sure if those would be accessible to this block once we merge a menus endpoint (@wpscholar has put together [a patch that covers all CRUD operations on menus](https://core.trac.wordpress.org/attachment/ticket/40878/40878.4.diff))

Even if a nav menu block is saved as a reusable block, this PR would basically store everything for that menu as a blob inside of `post_content`. This makes it much more difficult to access and parse from an API perspective. 

It also means we'd have to have 2 different ways to return menus via the REST API depending on how they were created. 

The current storage method could still allow rich block-based menus. Since individual menu items are already stored as posts, their extra block markup could be stored in `post_content`. I think the main thing that would need to be figured out is how to save any settings that apply to the menu as a whole.

**Comment by youknowriad** [comment]
@earnjam I think we should distinguish the menu block from the current menus. In the block editor, blocks have their data as attributes, blocks can be used in several contexts, inside post_content, soon maybe inside widget areas, and potentially later in "global templates". Blocks can also be used in a WP agnostic context.

For this reason, I think the BC concern is outside of the scope of this PR and there's no other way. The block is going to be used like any other block and should work in any of these contexts. Whether this block can be used in the menu screen and its attributes mapped to the existing storage is a separate question that is not impacted by this PR.

**Comment by earnjam** [comment]
@youknowriad I understand the desire for a platform agnostic menu block, but perhaps that block shouldn't be the default menu block for WordPress. 

Because that means we're either excluding traditional WordPress menus from being placed into areas that blocks can be used, or we'll require another "legacy menu" block that allows them to be inserted, which is just confusing. 

That also means that for the menus of the future, we're replacing a more structured storage implementation with one that is more difficult to parse. If you want to get menus via an API endpoint for some external client, then there will be multiple ways that they need to be retrieved and/or parsed.

**Comment by youknowriad** [comment]
I don't think we should base our future decision on what the WP menu block should be on the fact that WordPress has already existing menu concept. We should be thinking about what is the best menu block we could build for the future and then try to adapt it to work in WordPress.

> Because that means we're either excluding traditional WordPress menus from being placed into areas that blocks can be used or we'll require another "legacy menu" block that allows them to be inserted, which is just confusing.

We don't need any other block, we already have the "Classic Widget" block than can be used for the existing menus.

> That also means that for the menus of the future, we're replacing a more structured storage implementation with one that is more difficult to parse. If you want to get menus via an API endpoint for some external client, then there will be multiple ways that they need to be retrieved and/or parsed.

Why do you think so? For the future, I expect APIs to retrieve blocks composed of attributes.

**Comment by earnjam** [comment]
> We don't need any other block, we already have the "Classic Widget" block than can be used for the existing menus.

That's going to be pretty confusing for users. They want to insert a menu and choose the menu block, but then can't use the menu that they've already created. It isn't intuitive at all that the Legacy Widget block should be used for inserting menus. 

> > That also means that for the menus of the future, we're replacing a more structured storage implementation with one that is more difficult to parse. If you want to get menus via an API endpoint for some external client, then there will be multiple ways that they need to be retrieved and/or parsed.
> 
> Why do you think so? For the future, I expect APIs to retrieve blocks composed of attributes.

Let's say a site has a traditional menu and a user creates a menu using this block. How would you retrieve both of those via the REST API? 

The API either needs to handle the logic of retrieving and parsing two types of data, or else there are two endpoints.

**Comment by sarahmonster** [comment]
> That's going to be pretty confusing for users. They want to insert a menu and choose the menu block, but then can't use the menu that they've already created. 

Agreed! This is why our original design proposal suggests the ability to, upon creating a new menu block, _convert your existing menus_ to blocks. 

Nav- Smart default decision tree

...where "User selects from existing menus" looks something like this:
Existing menu selection

**Comment by youknowriad** [comment]
> Let's say a site has a traditional menu and a user creates a menu using this block. How would you retrieve both of those via the REST API?

By asking the question, you assume "menus" are an important concept that needs to fetched using the REST API. This assumption is completely true today (if we omit some menus created in an adhoc way including plugins to create menus) but is this assumption valid for the future?

Potentially, In terms of content, there's mainly two/three concept that matters for the user:

 - CPTs (posts, pages...)
 - Blocks in the content of these CPTS
 - Templates assigned to these CPTs. 

Content creation happens through these-screen (editors) and based on blocks. Menus are less important and represent just another block. Third-party blocks can create custom menu blocks, custom menu items. In fact, a "Recent Post" block is not much different from a menu, it's just a menu composed of given posts/pages... A list of buttons/links aligned next to each other in templates are also menus, what matters to the user at the point is the fact that the links work as expected and that they are global and edited in a single place. 

What is a menu in that case? Can you fetch recent post menu items today? Is it important for the user at that point to fetch menu items of a given block but not the other? I think it's more important for people to fetch blocks that compose templates and blocks that compose post_content at that time.

To be honest, I think we don't know the all answers to these questions. I don't think we should assume things and adapt the block editor's framework for something that we can't know for sure. 

That's why I suggest that we build the best menu block we can think of UI-wise first and iteration will inform us about the best storage system, how do users would want to consume the blocks...

**Comment by earnjam** [comment]
Thanks @youknowriad I don't want to be argumentative here. I appreciate you taking the time to outline your thoughts. 

I just want to make sure to voice some backward compatibility concerns. But I guess we are taking the same sort of mindset as the editor rollout. i.e. At a certain point we replace the menu screen entirely and people can just convert existing menus to new block-based menus when they go to edit them.

**Comment by youknowriad** [comment]
Thanks @earnjam for the understanding. I definitely don't want to dismiss any feedback and BC is still an important concern. It's important that we keep iterating in the plugin which will allow us to discover the shortcomings, the BC concerns and the use-cases that we may not cover properly and we'll have to adapt as these get clarified.

**Comment by sarahmonster** [comment]
Hello everyone! @jorgefilipecosta has informed me we can merge this PR without committing to a particular direction in terms of how the menu items are stored for now, and progress on the development of this block is still stalled pending getting this merged. 

Let's continue the ongoing debate, but if someone has time to review this PR and get it over the finish line, we'd really appreciate that. ❤️

**Comment by spacedmonkey** [comment]
Is it know theme creators will need to do support this new block? The css can get pretty complex for menus and menus are likely to look extremeley different depending on where there are on the site. Menus in a footer would be different from in content or sidebar for example. Is there going to be a way for developers to opt in to this block? Automatically opting them in will like end up with broken blocks on the front end.

**Comment by jorgefilipecosta** [comment]
> Is it know theme creators will need to do support this new block? The css can get pretty complex for menus and menus are likely to look extremeley different depending on where there are on the site. Menus in a footer would be different from in content or sidebar for example. Is there going to be a way for developers to opt in to this block? Automatically opting them in will like end up with broken blocks on the front end.

Hi, thank you for bringing these points. Right now the plan is to have a standard block and polish its UI, making improvements in the process ( to InnerBlocks for example) that any block developer can take advantage. There are ways themes can use to disable a specific block, regarding the styles the block will provide a default set of styles like other blocks do, and themes can opt-in to remove some of this styles. 



This PR was rebased taking into account updates that happen meanwhile, and the changes in https://github.com/WordPress/gutenberg/pull/14867, it should be ready to merge after https://github.com/WordPress/gutenberg/pull/14867.

**Review by draganescu — CHANGES_REQUESTED** [review]
I have tested locally and the blocks work in this alpha state. It works as described! Also the code itself is a stub for a lot more functionality to be added so we'll see what the final form will be. 

Committing this hidden away is a step forward so we can continue with smaller, more focused PRs for each part of the navigation menu block.

@jorgefilipecosta there is though a failing "Filter merged" check , not sure yet what this is about, never seen that before, so if you could let me know or fix it we can move along 💃

**Comment by jorgefilipecosta** [comment]
Hi @andreilupu, I rebased this PR, there are no errors happening now. Before merging this PR we need to merge https://github.com/WordPress/gutenberg/pull/14867, very simple PR just passing down a prop.

**Comment by andreilupu** [comment]
> Hi @andreilupu, I rebased this PR, there are no errors happening now. Before merging this PR we need to merge #14867, very simple PR just passing down a prop.

As much as I'm honored about the @ reference, I think that the shout out was for @draganescu 

I can still test the PR later if it's needed.

**Review by draganescu — APPROVED** [review]

**Pull request merged**
